### PR TITLE
Move Find All References to LSP

### PR DESCRIPTION
**Pull request by allisonchou**
Addresses #38477 and #42581. UI thread dependencies have been removed for the LSP scenario but kept for the local scenario.

Two items of note -
1) Currently, some parts of FAR do not display correctly: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1088938.

2) In the future, once Rich-Nav is implemented, the 'Origin' property in VSReferenceItem should be populated: https://github.com/dotnet/roslyn/issues/42847

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
1. Can you show a picture of what the results look like?
2. does this impact normal users, or only nexus users?

File: src/EditorFeatures/Core/FindUsages/AbstractFindUsagesService.cs [hunk 1]
```diff
@@ -88,6 +88,14 @@ protected AbstractFindUsagesService(IThreadingContext threadingContext)
             }
         }
 
+        async Task IFindUsagesLSPService.FindReferencesAsync(
+            Document document, int position, IFindUsagesContext context)
+        {
+            var definitionTrackingContext = new DefinitionTrackingContext(context);
+            await FindLiteralOrSymbolReferencesAsync(
+                document, position, definitionTrackingContext).ConfigureAwait(false);
+        }
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/FindUsages/AbstractFindUsagesService.cs]: you might want to doc the difference here.

File: src/EditorFeatures/Core/FindUsages/IFindUsagesLSPService.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.FindUsages;
+using Microsoft.CodeAnalysis.Host;
+
+namespace Microsoft.CodeAnalysis.Editor.FindUsages
+{
+    internal interface IFindUsagesLSPService : IFindUsagesService
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/FindUsages/IFindUsagesLSPService.cs]: is there value in subclassing here.  i'm not seeing it.

File: src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs [hunk 3]
```diff
@@ -372,5 +374,79 @@ public static Glyph CompletionItemKindToGlyph(LSP.CompletionItemKind kind)
...
+                }
+
+                if ((usageInfo & TypeOrNamespaceUsageInfo.Import) != 0)
+                {
+                    referenceKinds.Add(LSP.ReferenceKind.Import);
+                }
+
+                if ((usageInfo & TypeOrNamespaceUsageInfo.NamespaceDeclaration) != 0)
+                {
+                    referenceKinds.Add(LSP.ReferenceKind.Declaration);
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs]: for places where tehse don't align, can we comment why the mapping is felt to be correct?

File: src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs [hunk 4]
```diff
@@ -372,5 +374,79 @@ public static Glyph CompletionItemKindToGlyph(LSP.CompletionItemKind kind)
...
+                }
+            }
+
+            if (referenceKinds.IsEmpty())
+            {
+                referenceKinds.Add(LSP.ReferenceKind.Inactive);
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs]: what does this mean? (doc please).

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        {
+            _threadingContext = threadingContext;
+        }
+
+        public async Task<object[]> HandleRequestAsync(Solution solution, ReferenceParams referenceParams, ClientCapabilities clientCapabilities, CancellationToken cancellationToken)
+        {
+            // The VS LSP client supports streaming using IProgress<T> on various requests.
+            // However, this is not yet supported through Live Share, so deserialization fails on the IProgress<T> property.
+            // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1043376 tracks Live Share support for this (committed for 16.6).
+            var locations = ArrayBuilder<LSP.Location>.GetInstance();
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs]:
```suggestion
            using var _ = ArrayBuilder<LSP.Location>.GetInstance(out var locations);
```

Tihs way we don't leak if canceled.  Note: call .ToArray below, not ToArrayAndFree

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+            var document = solution.GetDocumentFromURI(referenceParams.TextDocument.Uri);
+            if (document == null)
+            {
+                return locations.ToArrayAndFree();
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs]: these sorts of things worry me.  it seems like it indicates a bug, but we silently fail instead.  i'm not sure how  auser would tell what's going on.

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+            }
+
+            var findUsagesService = document.Project.LanguageServices.GetService<IFindUsagesLSPService>();
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs]:
```suggestion
            var findUsagesService = document.GetLanguageService<IFindUsagesLSPService>();
```

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+            var position = await document.GetPositionFromLinePositionAsync(ProtocolConversions.PositionToLinePosition(referenceParams.Position), cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs]: wrap crazy lines :)

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+            var context = new SimpleFindUsagesContext(cancellationToken);
+
+            await findUsagesService.FindReferencesAsync(document, position, context).ConfigureAwait(false);
+
+            if (clientCapabilities?.HasVisualStudioLspCapability() == true)
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs]: doc?

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+                {
+                    definitionMap.Add(reference.Definition, new List<SourceReferenceItem>());
+                }
+
+                definitionMap[reference.Definition].Add(reference);
+            }
+
+            // NOTE: Parts of FAR currently do not display correctly due to a bug in LSP.VSReferenceItem.
+            // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1088938/
+            var referenceItems = ArrayBuilder<LSP.VSReferenceItem>.GetInstance();
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs]: use `using` pattern here.

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+
+            // Each reference is assigned its own unique id
+            var id = 0;
+            foreach (var definitionAndReferencesPair in definitionMap)
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs]:
```suggestion
            foreach (var (definition, references) in definitionMap)
```

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+            {
+                // Creating the reference item that corresponds to the definition
+                var definition = definitionAndReferencesPair.Key;
+                var definitionItem = await GenerateReferenceItem(id, definitionId: id, definition.SourceSpans.FirstOrDefault(), context, definition.DisplayableProperties,
+                    cancellationToken, definition.GetClassifiedText()).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs]: wrap better plz :)   should be visible on github.

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+
+                var definitionId = id;
+                id++;
+
+                // Creating a reference item for each reference
+                var references = definitionAndReferencesPair.Value;
+                foreach (var reference in references)
+                {
+                    var referenceItem = await GenerateReferenceItem(id, definitionId, reference.SourceSpan, context, reference.AdditionalProperties, cancellationToken,
+                        symbolUsageInfo: reference.SymbolUsageInfo).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs]: cancellationToken should be last.

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+                    ProjectName = documentSpan.Document.Project.Name,
+                    ResolutionStatus = ResolutionStatusKind.ConfirmedAsReference,
+                    Text = new ClassifiedTextElement(classifiedSpans.Select(cspan => new ClassifiedTextRun(cspan.ClassificationType, docText.ToString(cspan.TextSpan)))),
+                };
+            }
+        }
+
+        private static async Task<LSP.Location[]> GetLocationsAsync(LSP.ReferenceParams request, SimpleFindUsagesContext context, CancellationToken cancellationToken)
+        {
+            var locations = ArrayBuilder<LSP.Location>.GetInstance();
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs]: use using spattern.

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+
+            await findUsagesService.FindReferencesAsync(document, position, context).ConfigureAwait(false);
+
+            if (clientCapabilities?.HasVisualStudioLspCapability() == true)
+            {
+                return await GetReferenceItemsAsync(referenceParams, context, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                return await GetLocationsAsync(referenceParams, context, cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs]: is it actually important to support this case?

File: src/Features/LanguageServer/Protocol/LanguageServerProtocol.cs [hunk 16]
```diff
@@ -225,18 +237,6 @@ public Task<LSP.SymbolInformation[]> GetWorkspaceSymbolsAsync(Solution solution,
...
-        /// Answers a rename request by returning the workspace edit for a given symbol.
-        /// https://microsoft.github.io/language-server-protocol/specification#textDocument_rename
-        /// </summary>
-        /// <param name="solution">the solution containing the request.</param>
-        /// <param name="request">the document position of the symbol to rename.</param>
-        /// <param name="clientCapabilities">the client capabilities for the request.</param>
-        /// <param name="cancellationToken">a cancellation token.</param>
-        /// <returns>the workspace edits to rename the given symbol</returns>
-        public Task<WorkspaceEdit> RenameAsync(Solution solution, LSP.RenameParams request, LSP.ClientCapabilities clientCapabilities, CancellationToken cancellationToken)
-            => ExecuteRequestAsync<LSP.RenameParams, WorkspaceEdit>(LSP.Methods.TextDocumentRenameName, solution, request, clientCapabilities, cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/LanguageServerProtocol.cs]: revert?

File: src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs [hunk 17]
```diff
@@ -116,10 +116,17 @@ public void Exit()
         [JsonRpcMethod(Methods.TextDocumentRenameName)]
         public Task<WorkspaceEdit> GetTextDocumentRenameAsync(JToken input, CancellationToken cancellationToken)
         {
-            var renameParams = input.ToObject<RenameParams>();
+            var renameParams = input.ToObject<RenameParams>(JsonSerializer);
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs]: why this change?

**Comment by CyrusNajmabadi** [comment]
Overall looking good!  Mostly cleanup concerns.

**Comment by allisonchou** [comment]
> 1. Can you show a picture of what the results look like?
> 2. does this impact normal users, or only nexus users?

1. The results look like this today, with the first entry here being the definition and the second being a reference:
FAR
2. This impacts just LiveShare and Nexus users.

**Comment by CyrusNajmabadi** [comment]
What does this mean:

image

?

File: src/Features/LanguageServer/Protocol/LanguageServerProtocol.cs (hunk 16 above)
Comment (allisonchou) [reply src/Features/LanguageServer/Protocol/LanguageServerProtocol.cs]: I added this method in a previous PR but noticed in this PR that the rest of the method names were arranged alphabetically, so I just moved it for consistency.

**Comment by allisonchou** [comment]
> What does this mean:
> 
> image
> 
> ?
Currently, ClassifiedTextElement (the definition text type) isn't being recognized as a valid type for definitions, so it's being displayed incorrectly. This has been reported and I believe a fix is underway.

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs (hunk 6 above)
Comment (allisonchou) [reply src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs]: Hmm... do you think calling .GetRequiredDocument instead would be better? But since it would throw if the document can't be found, I'm not sure that would be an improvement...

File: src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs (hunk 17 above)
Comment (allisonchou) [reply src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs]: I believe it's a workaround for the issue mentioned here: https://github.com/dotnet/roslyn/blob/master-vs-deps/src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs#L35
All other methods seem to have been modified to fit this change, I think since rename was a recent addition, it may not have been initially included when the modifications were done.

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs (hunk 15 above)
Comment (allisonchou) [reply src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs]: After talking with David, it was apparently included in order to support Liveshare backwards compatibility, but since back-compat is already broken, I was told it's probably ok to remove for now if an assert was included.

File: src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs:441]: you resolved... but no comment wer eadded to the code :)

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.Text.Adornments;
+using Roslyn.Utilities;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler
+{
+    // The VS LSP client supports streaming using IProgress<T> on various requests.
+    // However, this is not yet supported through Live Share, so deserialization fails on the IProgress<T> property.
+    // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1043376 tracks Live Share support for this (committed for 16.6).
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs:33]: @dibarbet @allisonchou aren't we in 16.6?  does this ocmment still apply?

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs [hunk 19]
```diff
@@ -0,0 +1,183 @@
...
+            Solution solution,
+            ReferenceParams referenceParams,
+            ClientCapabilities clientCapabilities,
+            CancellationToken cancellationToken)
+        {
+            Debug.Assert(clientCapabilities.HasVisualStudioLspCapability());
+
+            // The VS LSP client supports streaming using IProgress<T> on various requests.
+            // However, this is not yet supported through Live Share, so deserialization fails on the IProgress<T> property.
+            // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1043376 tracks Live Share support for this (committed for 16.6).
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs:56]: duplicate comment from above.  not sure if we need both.

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs (hunk 18 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs:33]: Yep! this should be fixed in master-vs-deps (it requires 16.6 package versions).  This comment should no longer apply.

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs [hunk 20]
```diff
@@ -0,0 +1,183 @@
...
+            var document = solution.GetDocumentFromURI(referenceParams.TextDocument.Uri);
+            if (document == null)
+            {
+                return new LSP.VSReferenceItem[] { };
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs:60]: `Array.Empty<LSP.VSReferenceItem>();` is preferred idiom.

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs [hunk 21]
```diff
@@ -0,0 +1,183 @@
...
+            // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1088938/
+            using var _ = ArrayBuilder<LSP.VSReferenceItem>.GetInstance(out var referenceItems);
+
+            // Each reference is assigned its own unique id
+            var id = 0;
+            foreach (var (definition, references) in definitionMap)
+            {
+                // Creating the reference item that corresponds to the definition
+                var definitionItem = await GenerateReferenceItem(
+                    id, definitionId: id, definition.SourceSpans.FirstOrDefault(), context, definition.DisplayableProperties,
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs:101]: small problem.  what if the definition is in metadata and has no source-spans.  does this work properly?  can you manually test that a metadata-find-refs still has proper results that can be grouped/navigated.  Also, ensure we don't crash if the user tries to navigate to the definition itself

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs [hunk 22]
```diff
@@ -0,0 +1,183 @@
...
+                DocumentSpan documentSpan,
+                SimpleFindUsagesContext context,
+                ImmutableDictionary<string, string> properties,
+                CancellationToken cancellationToken,
+                ClassifiedTextElement? definitionText = null,
+                SymbolUsageInfo? symbolUsageInfo = null)
+            {
+                var location = await ProtocolConversions.DocumentSpanToLocationAsync(documentSpan, cancellationToken).ConfigureAwait(false);
+                var classifiedSpansAndHighlightSpan = await ClassifiedSpansAndHighlightSpanFactory.ClassifyAsync(
+                    documentSpan, context.CancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs:133]:
```suggestion
                    documentSpan, cancellationToken).ConfigureAwait(false);
```

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs [hunk 23]
```diff
@@ -0,0 +1,183 @@
...
+                var classifiedSpans = classifiedSpansAndHighlightSpan.ClassifiedSpans;
+                var docText = await documentSpan.Document.GetTextAsync(context.CancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs:135]:
```suggestion
                var docText = await documentSpan.Document.GetTextAsync(cancellationToken).ConfigureAwait(false);
```

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs [hunk 24]
```diff
@@ -0,0 +1,183 @@
...
+
+            static async Task<LSP.VSReferenceItem> GenerateReferenceItem(
+                int id,
+                int? definitionId,
+                DocumentSpan documentSpan,
+                SimpleFindUsagesContext context,
+                ImmutableDictionary<string, string> properties,
+                CancellationToken cancellationToken,
+                ClassifiedTextElement? definitionText = null,
+                SymbolUsageInfo? symbolUsageInfo = null)
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs:129]: i would make these non-optiona.  then have the caller be explicit at they're intending to provide 'null' here as they're not relevant.

in general optional parameters for internal APIs are *highly* discouraged.

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs (hunk 14 above)
Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs:162]:
```suggestion
            using var _ = ArrayBuilder<LSP.Location>.GetInstance(out var locations);
```

That's the pattern i meeant :)

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs [hunk 25]
```diff
@@ -0,0 +1,183 @@
...
+                    }
+                }
+            }
+
+            foreach (var reference in context.GetReferences())
+            {
+                locations.Add(await ProtocolConversions.DocumentSpanToLocationAsync(reference.SourceSpan, cancellationToken).ConfigureAwait(false));
+            }
+
+            return locations.ToArrayAndFree();
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs:180]:
```suggestion
            return locations.ToArray();
```

File: src/Features/LanguageServer/ProtocolUnitTests/References/FindAllReferencesHandlerTests.cs [hunk 26]
```diff
@@ -37,34 +35,8 @@ void M2()
...
-    }
-}
-class B
-{
-    int someInt = A.{|reference:someInt|} + 1;
-    void M2()
-    {
-        var j = someInt + A.{|caret:|}{|reference:someInt|};
-    }
-}";
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/ProtocolUnitTests/References/FindAllReferencesHandlerTests.cs:63]: why was this test removed?

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs (hunk 18 above)
Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs:33]: Can you walk @allisonchou through how to use IProgress then?  (or point her at an example that does that)?

File: src/Features/LanguageServer/ProtocolUnitTests/References/FindAllReferencesHandlerTests.cs (hunk 26 above)
Comment (allisonchou) [reply src/Features/LanguageServer/ProtocolUnitTests/References/FindAllReferencesHandlerTests.cs:63]: In Roslyn's local FAR experience, I didn't see an option to not include definitions. (I might be totally off base here, feel free to correct me if so.) Additionally, in the new ReferenceItem class, I believe it works by associating every reference with a definition, so I'm not sure if it's an option to leave out the definition (at least on our end).

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs (hunk 21 above)
Comment (allisonchou) [reply src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs]: This was a great catch, the code didn't do this properly before, but I believe it should be fixed now. Thanks!

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs (hunk 18 above)
Comment (dibarbet) [reply src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs]: Sure - I believe the plan was to first move this to the actual LSP client, then do a followup with the streaming changes.  felt that it might be easier to split them up to reduce the complexity (in writing and reviewing)

@allisonchou ping me if you have any questions.  The current example we have is here - https://github.com/dotnet/roslyn/blob/master-vs-deps/src/Workspaces/Remote/ServiceHub/Services/LanguageServer.cs#L125 which we'd probably need to hook up to the findusages context

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs [hunk 27]
```diff
@@ -0,0 +1,195 @@
...
+            {
+                LSP.Location? location = null;
+
+                // If we have no source span, our location may be in metadata.
+                if (documentSpan == default)
+                {
+                    var symbol = await SymbolFinder.FindSymbolAtPositionAsync(originalDocument, originalPosition, cancellationToken).ConfigureAwait(false);
+                    if (symbol != null && symbol.Locations != null && !symbol.Locations.IsEmpty && symbol.Locations.First().IsInMetadata)
+                    {
+                        var declarationFile = await _metadataAsSourceFileService.GetGeneratedFileAsync(originalDocument.Project, symbol, false, cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs:141]: 1. named parmaeter for 'false'.  
2. does this actually work?  if so... nifty.

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs [hunk 28]
```diff
@@ -0,0 +1,195 @@
...
+
+                        var linePosSpan = declarationFile.IdentifierLocation.GetLineSpan().Span;
+                        location = new LSP.Location
+                        {
+                            Uri = new Uri(declarationFile.FilePath),
+                            Range = ProtocolConversions.LinePositionToRange(linePosSpan),
+                        };
+                    }
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs:149]: what happens if we get here and haven't set hte location?  what is the experience then?

File: src/Features/LanguageServer/ProtocolUnitTests/References/FindAllReferencesHandlerTests.cs (hunk 26 above)
Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/ProtocolUnitTests/References/FindAllReferencesHandlerTests.cs:63]: in general, try not to remove tests.  instead, just update the test to the new 'expected' behavior.

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs (hunk 18 above)
Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs]: i'm find with that.  let's get this in asap, then we can do streaming :)

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs (hunk 28 above)
Comment (allisonchou) [reply src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs:149]: Hmm, from testing, it looks like if the location property is empty, no references show up at all. Will push a change to skip definitions and references without locations.

File: src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs (hunk 27 above)
Comment (allisonchou) [reply src/Features/LanguageServer/Protocol/Handler/References/FindAllReferencesHandler.cs:141]: From testing it does seem to work, I based the code off similar metadata code in the GoToDef handler.

File: src/Features/LanguageServer/ProtocolUnitTests/References/FindAllReferencesHandlerTests.cs (hunk 26 above)
Comment (allisonchou) [reply src/Features/LanguageServer/ProtocolUnitTests/References/FindAllReferencesHandlerTests.cs:63]: Gotcha. For this test in particular, I believe the expected behavior is identical to the test right above it, so I don't know if there would be value in keeping it around.

Comment (CyrusNajmabadi) [reply src/Features/LanguageServer/ProtocolUnitTests/References/FindAllReferencesHandlerTests.cs:63]: good to know. that's an ok reason to remove.  if the test is identical though, please comment as such in the PR so that's known :)
